### PR TITLE
Implement F4Jumble

### LINKED ISSUE #378: Implement F4Jumble
Required for Unified Addresses.

### PR DESCRIPTION
**Pull request by nuttycom**
Fixes #378

### CODE REVIEW

File: zcash_primitives/src/address/f4jumble.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
+use blake2b_simd::{Params as Blake2bParams, OUTBYTES};
+use std::cmp::min;
```
Comment (str4d) [inline zcash_primitives/src/address/f4jumble.rs:2]:
```suggestion
use std::cmp::min;
use std::ops::RangeInclusive;

const VALID_LENGTH: RangeInclusive<usize> = 48..=16448;
```

File: zcash_primitives/src/address/f4jumble.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+fn xor(a: &[u8], b: &[u8]) -> Vec<u8> {
+    a.iter().zip(b.iter()).map(|(a0, b0)| a0 ^ b0).collect()
+}
+
+fn ceildiv(num: usize, den: usize) -> usize {
+    (num + den - 1) / den
+}
+
+#[allow(clippy::many_single_char_names)]
+pub fn f4jumble(mut a: Vec<u8>) -> Option<Vec<u8>> {
```
Comment (str4d) [inline zcash_primitives/src/address/f4jumble.rs:61]:
```suggestion
pub fn f4jumble(a: &[u8]) -> Option<Vec<u8>> {
```

File: zcash_primitives/src/address/f4jumble.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            c.extend(d);
+            c
+        })
+    } else {
+        None
+    }
+}
+
+#[allow(clippy::many_single_char_names)]
+pub fn f4jumble_inv(mut c: Vec<u8>) -> Option<Vec<u8>> {
```
Comment (str4d) [inline zcash_primitives/src/address/f4jumble.rs:81]:
```suggestion
pub fn f4jumble_inv(c: &[u8]) -> Option<Vec<u8>> {
```

File: zcash_primitives/src/address/f4jumble.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+mod tests {
+    use proptest::collection::vec;
+    use proptest::prelude::*;
+
+    use super::{f4jumble, f4jumble_inv};
+
+    proptest! {
+        #[test]
+        fn f4jumble_roundtrip(msg in vec(any::<u8>(), 48..16448)) {
+            let jumbled = f4jumble(msg.clone()).unwrap();
```
Comment (str4d) [inline zcash_primitives/src/address/f4jumble.rs:110]:
```suggestion
            let jumbled = f4jumble(&msg).unwrap();
```

File: zcash_primitives/src/address/f4jumble.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            let jumbled_len = jumbled.len();
+            prop_assert_eq!(
+                msg.len(), jumbled_len,
+                "Jumbled length {} was not equal to message length {}",
+                jumbled_len, msg.len()
+            );
+
+            let unjumbled = f4jumble_inv(jumbled).unwrap();
```
Comment (str4d) [inline zcash_primitives/src/address/f4jumble.rs:118]:
```suggestion
            let unjumbled = f4jumble_inv(&jumbled).unwrap();
```

**Review by str4d — CHANGES_REQUESTED** [review]

**Comment by str4d** [comment]
ALso, add tests that run this against the test vectors from https://github.com/zcash-hackworks/zcash-test-vectors/pull/31 (once that PR is merged).

File: zcash_primitives/src/address/f4jumble.rs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            result.extend(
+                Blake2bParams::new()
+                    .hash_length(OUTBYTES)
+                    .personal(&personal)
+                    .hash(u)
+                    .as_ref(),
+            );
+        }
+
+        result.into_iter().take(self.l_r).collect()
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_primitives/src/address/f4jumble.rs:48]: I tried everything to get this to be a flat_map, I'll have to see if this compiles.

[the remark this answers is not in the document]
Comment (str4d) [reply zcash_primitives/src/address/f4jumble.rs:48]: As long as the output can be moved it should be fine; `.to_vec().into_iter()` should achieve this.

File: zcash_primitives/src/address/f4jumble.rs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+            prop_assert_eq!(msg, unjumbled, "Unjumbled message did not match original message.");
+        }
+    }
+
+    #[test]
+    fn f4jumble_check_vectors() {
+        for v in test_vectors() {
+            let jumbled = f4jumble(&v.normal).unwrap();
+            assert_eq!(jumbled, v.jumbled);
```
Comment (str4d) [inline zcash_primitives/src/address/f4jumble.rs:150]: Also test the other direction:
```suggestion
            let jumbled = f4jumble(&v.normal).unwrap();
            assert_eq!(jumbled, v.jumbled);
            let unjumbled = f4jumble_inv(&v.jumbled).unwrap();
            assert_eq!(unjumbled, v.normal);
```

**Review by str4d — APPROVED** [review]
utACK with @daira and @therealyingtong.

**Pull request merged**